File: contexts/python/services/coding_agent.py
"""Coding agent — given an incident and a sandbox, drafts a fix and runs the
app's tests (the smoke gate). `CannedCodingAgent` recognizes a small set of
seeded bugs and applies the known fix. A real agent investigates and writes the
patch itself — implement `CLICodingAgent` (ESSENCE §4E, §6.2)."""
from dataclasses import dataclass, field

from services._fixes import CANNED_FIXES


@dataclass
class FixResult:
    status: str  # succeeded | smoke_failed | diff_empty | deferred
    diff: str = ""
    files: dict = field(default_factory=dict)
    smoke_passed: bool = False
    smoke_output: str = ""
    narrative: str = ""


class CannedCodingAgent:
    def attempt_fix(self, incident, sandbox) -> FixResult:
        fix = CANNED_FIXES.get(incident.slug)
        if fix is None:
            return FixResult(
                status="deferred",
                narrative=f"No canned fix for '{incident.slug}'. A real coding agent would write one.",
            )
        sandbox.write_files(fix)
        passed, out = sandbox.run_tests()
        diff = sandbox.diff()
        if not diff:
            return FixResult(status="diff_empty", smoke_output=out, narrative="no change produced")
        return FixResult(
            status="succeeded" if passed else "smoke_failed",
            diff=diff, files=fix, smoke_passed=passed, smoke_output=out,
            narrative="applied the fix and ran the app's tests in the sandbox",
        )


class CLICodingAgent:
    """REAL coding-agent adapter — IMPLEMENT ME (ESSENCE §4E, §6.2). Run your
    coding CLI against the sandbox with the incident inlined; return diff + smoke."""

    def attempt_fix(self, incident, sandbox) -> FixResult:
        raise NotImplementedError("Wire your coding agent — ESSENCE.md §6.2.")

File: contexts/typescript/services/coding_agent.ts
// Coding agent — given an incident and a sandbox, drafts a fix and runs the
// app's tests (the smoke gate). `CannedCodingAgent` recognizes a small set of
// seeded bugs and applies the known fix. A real agent investigates and writes
// the patch itself — implement `CLICodingAgent` (ESSENCE §4E, §6.2).
import { CANNED_FIXES } from "./fixes.ts";
import type { Incident } from "./triage_model.ts";
import type { LocalSandbox } from "./sandbox.ts";

export interface FixResult {
  status: string; // succeeded | smoke_failed | diff_empty | deferred
  diff: string;
  files: Record<string, string>;
  smokePassed: boolean;
  smokeOutput: string;
  narrative: string;
}

export class CannedCodingAgent {
  attemptFix(incident: Incident, sandbox: LocalSandbox): FixResult {
    const fix = CANNED_FIXES[incident.slug];
    if (!fix) {
      return {
        status: "deferred", diff: "", files: {}, smokePassed: false, smokeOutput: "",
        narrative: `No canned fix for '${incident.slug}'. A real coding agent would write one.`,
      };
    }
    sandbox.writeFiles(fix);
    const { passed, output } = sandbox.runTests();
    const diff = sandbox.diff();
    if (!diff) {
      return { status: "diff_empty", diff: "", files: {}, smokePassed: passed, smokeOutput: output, narrative: "no change produced" };
    }
    return {
      status: passed ? "succeeded" : "smoke_failed",
      diff, files: fix, smokePassed: passed, smokeOutput: output,
      narrative: "applied the fix and ran the app's tests in the sandbox",
    };
  }
}

export class CLICodingAgent {
  attemptFix(_incident: Incident, _sandbox: LocalSandbox): FixResult {
    throw new Error("Wire your coding agent — ESSENCE.md §6.2.");
  }
}
